src/smart-components/role/add-role-new/cost-resources.js

```diff
@@ -0,0 +1,171 @@
+        ...state,
+        [action.key]: {
+          ...prevState,
+          selected: prevState.options.map((option) => option.value),
+        },
+      };
+    case 'clear':
+      return {
+        ...state,
+        [action.key]: {
+          ...prevState,
+          selected: [],
+        },
+      };
+    case 'setOptions':
+      return {
+        ...state,
+        [action.key]: {
+          ...prevState,
+          options: action.options,
+        },
+      };
+    default:
+      return state;
+  }
+};
+
+const CostResources = (props) => {
+  const dispatch = useDispatch();
+  const fetchData = (apiProps) => dispatch(getResourceDefinitions(apiProps));
+  const fetchResource = (apiProps) => dispatch(getResource(apiProps));
+  const { resourceTypes, isLoading, isLoadingResources, resources } = useSelector(selector, shallowEqual);
+  const { input } = useFieldApi(props);
+  const formOptions = useFormApi();
+  const { 'add-permissions-table': permissions } = formOptions.getState().values;
+  const [state, dispatchLocaly] = useReducer(
+    reducer,
+    permissions.reduce(
+      (acc, permission) => ({
+        ...acc,
+        [permission.uuid]: {
+          selected: [],
+          options: [],
+          isOpen: false,
+        },
+      }),
+      {}
+    )
+  );
+  const onToggle = (key, isOpen) => dispatchLocaly({ type: 'toggle', key, isOpen });
+  const clearSelection = (key) => dispatchLocaly({ type: 'clear', key });
+  const onSelect = (event, selection, selectAll, key) =>
+    selectAll ? dispatchLocaly({ type: 'selectAll', selection, key }) : dispatchLocaly({ type: 'select', selection, key });
+
+  const permissionToResource = (permission) => resourceTypes.find((r) => r.value === permission.split(':')[1])?.path.split('/')[5];
+
+  useEffect(() => {
+    fetchData();
+  }, []);
+
+  useEffect(() => {
+    if (!isLoading) {
+      const resourcePaths = [
+        ...new Set(permissions.map((permission) => resourceTypes.find((r) => r.value === permission.uuid.split(':')[1])?.path)),
+      ].filter((path) => path); // remove undefined
+      resourcePaths.map((path) => fetchResource(path));
+    }
+  }, [resourceTypes]);
+
+  useEffect(() => {
+    if (!isLoadingResources) {
+      permissions.map((p) => dispatchLocaly({ type: 'setOptions', key: p.uuid, options: resources[permissionToResource(p.uuid)] || [] }));
+    }
+  }, [isLoadingResources]);
+
+  useEffect(() => {
+    const resourceDefinitions = Object.entries(state).map(([permission, resources]) => ({ permission, resources: resources.selected }));
+    input.onChange(resourceDefinitions);
+    formOptions.change('resource-definitions', resourceDefinitions);
+  }, [state]);
+
+  // eslint-disable-next-line react/prop-types
+  const makeRow = ({ uuid: permission }) => {
+    const options = resources[permissionToResource(permission)] || [];
+    return (
+      <React.Fragment>
+        <GridItem span={3}>
+          <FormGroup label={permission} isRequired></FormGroup>
```

Comment:
How does this behave on small screens?
```jsx
<FormGroup label={permission} isRequired />
```

Reply:
I would say that it is ok
![Screenshot from 2020-10-13 10-05-40](https://user-images.githubusercontent.com/9535558/95833305-b7a1d500-0d3b-11eb-9f26-4c2fe77949da.png)

Reply:
I think it would be worthwhile to switch to a horizontal layout on small screens. It would definitely improve the mobile experience. The labels are just way too long.

Reply:
Yep, it is looks better. I have added screenshots to the description.

Reply:
@PanSpagetka @Hyperkid123  I think we need input from @mmenestr on what the mobile layout for this should look like. (I can't see repeating the labels on every line.) Anyway, this modal doesn't work in mobile at the moment anyway because the steps should switch to a dropdown in the header (Jira issue: https://issues.redhat.com/browse/RHCLOUD-9548)

 I say merge this and we'll deal with the mobile issues later.

Reply:
![Screen Shot 2020-10-13 at 9 55 07 AM](https://user-images.githubusercontent.com/1287144/95870231-41e93980-0d3a-11eb-970f-9cc555dd897c.png)